src/feature/dircommon/directory.c

```diff
@@ -167,6 +175,67 @@ connection_dir_is_encrypted(const dir_connection_t *conn)
   return TO_CONN(conn)->linked;
 }
 
+/** Return true iff the given directory connection <b>dir_conn</b> is
+ * anonymous, that is, it is on a circuit via a public relay and not directly
+ * from a client or bridge.
+ *
+ * For client circuits via relays: true for 2-hop+ paths.
+ * For client circuits via bridges: true for 3-hop+ paths.
+ *
+ * This first test if the connection is encrypted since it is a strong
+ * requirement for anonymity. */
+bool
+connection_dir_is_anonymous(const dir_connection_t *dir_conn)
+{
+  const connection_t *conn, *linked_conn;
+  const edge_connection_t *edge_conn;
+  const circuit_t *circ;
+
+  tor_assert(dir_conn);
+
+  if (!connection_dir_is_encrypted(dir_conn)) {
+    return false;
+  }
+
+  /*
+   * Buckle up, we'll do a deep dive into the connection in order to get the
+   * final connection channel of that connection in order to figure out if
+   * this is a client or relay link.
+   *
+   * We go: dir_conn -> linked_conn -> edge_conn -> on_circuit -> p_chan.
```

Comment:
I'm slightly afraid of this function giving a false negative and hence marking a connection as non-anonymous when it is, since that will have serious reachability issues for onion service. The problem is that the ladder from dir_conn -> p_chan is so complicated. Can we BUG out even more on the checks below if we consider them invariants, so that if this function does not work as intended we learn it soon enough?

Reply:
Sounds like a good idea.

In the original code, I had info logs. We should switch to info logs for the failures that actually happen.
https://github.com/teor2345/tor-old/commit/10290066c8ee6b4aa40ec048222fdd4f572ef8d9#diff-c56fd972333216da3bb1852bcc89f57dR1587